Remove TEST_CONFIG_FILE_KEY testing keyword

valentin-krasontovitsch opened the pull request:
Resolves #4229 

~~The config key `TEST_CONFIG_FILE_KEY` is only used in the context of tests, in test code. Writing it to the config file leads to warnings about an unknown config key. This change should suppress those warnings.~~

After a long discussion, with many good ideas, I have arrived at a different solution, please [see below](https://github.com/equinor/ert/pull/4215#issuecomment-1312809102)

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

valentin-krasontovitsch commented:
jenkins x-ert-ERT test failed with

```
tests/performance_tests/test_forward_model_script_overhead.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/rh/rh-python38/root/usr/lib64/python3.8/subprocess.py:359: in check_call
    retcode = call(*popenargs, **kwargs)
/opt/rh/rh-python38/root/usr/lib64/python3.8/subprocess.py:342: in call
    return p.wait(timeout=timeout)
/opt/rh/rh-python38/root/usr/lib64/python3.8/subprocess.py:1083: in wait
    return self._wait(timeout=timeout)
/opt/rh/rh-python38/root/usr/lib64/python3.8/subprocess.py:1806: in _wait
    (pid, sts) = self._try_wait(0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <subprocess.Popen object at 0x7fb9494f16a0>, wait_flags = 0

    def _try_wait(self, wait_flags):
        """All callers to this function MUST hold self._waitpid_lock."""
        try:
>           (pid, sts) = os.waitpid(self.pid, wait_flags)
E           Failed: Timeout >12.0s

/opt/rh/rh-python38/root/usr/lib64/python3.8/subprocess.py:1764: Failed
```

this seems to be a timeout / flaky thing.

jenkins, retest this please.

oysteoh commented:
Jenkins: Test this please!

oysteoh commented:
Sorry, but i'm not following totally what and why here :grimacing:  Do we at all need to add this key? If it is not supported, why do we need it? I guess if not supported there is no functionality around it... Seems weird to introduce code to avoid warning-messages if we can avoid it in the first place. But i probably miss a part of the picture here... :thinking:

oysteoh commented:
as i said - i might miss some important parts here - but what i'm thinking - why not do i.e this 

```
@pytest.mark.usefixtures("use_tmpdir")
@given(config_dicts())
def test_from_dict_and_from_file_creates_equal_subst_lists(config_dict):
    filename = "testconfig"
    to_config_file(filename, config_dict)
    res_config_from_file = ResConfig(user_config_file=filename)
    res_config_from_dict = ResConfig(config_dict=config_dict)
    assert (
        res_config_from_file.substitution_list == res_config_from_dict.substitution_list
    )
```

instead of what we have now
```
@pytest.mark.usefixtures("use_tmpdir")
@given(config_dicts())
def test_from_dict_and_from_file_creates_equal_subst_lists(config_dict):
    filename = config_dict[pytest.TEST_CONFIG_FILE_KEY]    
    to_config_file(filename, config_dict)
    res_config_from_file = ResConfig(user_config_file=filename)
    res_config_from_dict = ResConfig(config_dict=config_dict)
    assert (
        res_config_from_file.substitution_list == res_config_from_dict.substitution_list
    )
```

valentin-krasontovitsch commented:
during parsing, through the predefined keywords mechanisms, a couple of substitutions are done, one of them for the file name of the config file. these predefined keywords are later passed to the substitution list formerly known as substitution config.
this is in the config content path.

initializing res config from a config dict, we circumvent parsing, and create the dictionary by hand (generally using hypothesis). we still expect to find the substitutions that come from the predefined keywords mechanism of the config file parser. we have to inject them somehow! we hijack the define keyword, and put the substitutions there, during generation of the config dict. then we have to make sure that the config file name we put in the defines is also used for creation of the config file based on the config dict in the test. we have to make the test function aware of the config file name that we generated for the config dict. this is exactly what `pytest.TEST_CONFIG_FILE_KEY = __TEST_CONFIG_FILE` is used for.

sorry about the wall of text.

we will be able to get rid of this once we consolidate / get rid of config content and reimagine parsing.

xjules commented:
> during parsing, through the predefined keywords mechanisms, a couple of substitutions are done, one of them for the file name of the config file. these predefined keywords are later passed to the substitution list formerly known as substitution config. this is in the config content path.
> 
> initializing res config from a config dict, we circumvent parsing, and create the dictionary by hand (generally using hypothesis). we still expect to find the substitutions that come from the predefined keywords mechanism of the config file parser. we have to inject them somehow! we hijack the define keyword, and put the substitutions there, during generation of the config dict. then we have to make sure that the config file name we put in the defines is also used for creation of the config file based on the config dict in the test. we have to make the test function aware of the config file name that we generated for the config dict. this is exactly what `pytest.TEST_CONFIG_FILE_KEY = __TEST_CONFIG_FILE` is used for.
> 
> sorry about the wall of text.
> 
> we will be able to get rid of this once we consolidate / get rid of config content and reimagine parsing.

Started reading this out of curiosity of why the test fails but must say haven't understand the need for having the `TEST_CONFIG_FILE_KEY` neither unfortunately

oysteoh commented:
yes, it is funny a small change causing much discussion 😂  This is not your fault, it is ERT which is complicated and i really want to understand why this is i needed so i would be thankful to have an explanation before we merge! 🙏

valentin-krasontovitsch commented:
of course - let's have a look at it tomorrow then! the more people look at these complicated things, the more confidence we get in the changes we're doing at the moment, too!

oysteoh commented:
Still after looking into the original commit 841c599ec548464185b12578bd91d3e1645a3161 i cannot understand why we at all need this pytest key. Why not just remove it ? It is not testing any behavior is it? As i wrote further up, why not just do what i suggest in the tests there? Is it possible to explain? :grimacing:  https://github.com/equinor/ert/pull/4215#issuecomment-1308726995

eivindjahren commented on tests/test_config_parsing/config_dict_generator.py:
```diff
@@ -389,6 +389,8 @@ def to_config_file(filename, config_dict):  # pylint: disable=too-many-branches
             if keyword == ConfigKeys.DATA_KW_KEY:
                 for define_key, define_value in keyword_value.items():
                     config.write(f"{keyword} {define_key} {define_value}\n")
+            elif keyword == pytest.TEST_CONFIG_FILE_KEY:
```
I suggest we inline `pytest.TEST_CONFIG_FILE_KEY`and just write `keyword == "TEST_CONFIG_FILE"` everywhere.

valentin-krasontovitsch replied:
i'm not entirely sure what you mean - instead of holding the value in a variable, you want to repeat the string literal everywhere it's used?

eivindjahren replied:
Yes, that was generally the idea. Since it is just used in tests that seemed unproblematic.

eivindjahren replied:
Actually, I got a better idea, so I am closing this conversation.

valentin-krasontovitsch replied:
this is outdated now, the keyword is gone : )

eivindjahren commented:
The way I see it, we have not graduated from just generating a dictionary to write to file as we need to know more things from the generator. I might add that we have for a while had to mock the site config to get it to work (see `tests/test_config_parsing/contest.py`). So let's create a class that holds the relevant context of the config that we generate instead, ie. add the following class to `tests/test_config_parsing/config_dict_generator.py`

```python

@dataclass
class ErtConfig:
  config_dict : Dict[str, Any]
  config_file_name: str

  def write_config(self):
    ....
```

with corresponding generator `ert_configs` made using the composite strategy. Likely we have more context about the config that we want to add such as the location and contents of the site config. This class might also become part of the implementation later.

valentin-krasontovitsch commented:
OK so I have arrived at the following proposal, which aims to do two things:

- resolve the mess with the `TEST_CONFIG_FILE_KEY` constant, and
- move into a hopefully better direction with respect to the predefined defines

To that end, I have

- removed the highly controversial keyword altogether
  * by replacing it with a string literal `config.ert` in tests that don't care about the predefined defines, and
  * by using a new mechanism to make sure the config file name (and path) end up in the expected defines (in tests for res config's substitution list) [see below]
- removed the `CONFIG_DIRECTORY` key from our config dict generator based tests, as setting that one is not strictly speaking necessary (in all code paths where it's used, we default to cwd, which it was in the tests as well) [SIMPLIFY]
- created a new argument to res config, `config_dict_source_file`, which means to communicate some context about the config dict, and is used to create the predefined defines explicitly, which are again explicitly passed as an argument to the methods creating the substitution list (and it is made clear with a comment that the predefineds:tm:  are part of the defines in config content)

With this I hope to make clearer the distinction between content parsed from the config file (what ends up in config dict) and context content (location and name of the config file). This is moving towards more refined ideas of Morten and Eivind, but the predefineds are still just a free for all dict (sorry Morten), and the context info is not collected in a fancy dataclass (sorry Eivind) - I wanted to leave the changes here small, for now, with room for improvement later.

I hope this solution is more agreeable, but I have also emptied my calendar next week in case there is further need for discussions ; )

DanSava commented on src/ert/_c_wrappers/enkf/res_config.py:
```diff
@@ -104,7 +105,9 @@ def __init__(
             )
         else:
-            self._alloc_from_dict(config_dict=config_dict)
+            self._alloc_from_dict(
+                config_dict=config_dict, source_file=config_dict_source_file
+            )
```
This seems a bit confusing to me, this seems to suggest that we only have dictionaries that are read from a file. Is this the intended design?

eivindjahren replied:
I am wondering if this isn't less useful. Isn't it better that the list of defines is given explicitly?

valentin-krasontovitsch replied:
about dictionaries only read from file - it was my impression that there's where we moving towards / want to get in the end, and I did not consider other alternatives for getting a dictionary (besides from tests of course). do you have a more accurate naming suggestion, Dan?

Re: Eivind - we can totally change the interface and yank the creation of predefined defines out of resconfig, thanks for the suggestion! would you like me to let it live in...

so i'm gonna 
- change the interface of enkfmain, and let it accept a `predefined_defines: dict` argument, and
- move the convenience method for creating our current default to ...

hmm. i usually don't go higher up in the stack than enkf main, i see ResConfig is initialized

- in `__main__.py`,
- in `cli/main.py`, and
- in `libres_facade.py`

any suggestions as to where the method should live so it is accessible by all three?

eivindjahren replied:
What we have thought about is getting a `LibresFacade.from_dict` classmethod analogous to `LibresFacade.from_file` that hides the implementation detail of initializing `ResConfig` and `EnKFMain` mind that this opens up the pandoras box of moving everest away from initializing using `ResConfig(config=...)`.

DanSava replied:
> about dictionaries only read from file - it was my impression that there's where we moving towards / want to get in the end, and I did not consider other alternatives for getting a dictionary (besides from tests of course). do you have a more accurate naming suggestion, Dan?

Even though we are aiming at reading dictionaries from files we are not there yet and currently, there is no instance where a dictionary is read from a file (for ResConfig), so I am not sure it is a good idea to try and implement some distant future functionality now.

But let's say we want to try to do that I am not sure passing a dict and a file_name where that dict is being read from is a winning combination in any kind of future-proofing.  You would be either passing the filepath or the dict, if the implementation requires both then the implementation can be improved :)

My suggestion would be to focus on fixing the tests and not adding/changing functionality so that we can test it.

valentin-krasontovitsch replied:
i guess we can't really future proof at this stage, really I was just trying to move towards the right direction. at the moment, you're of course right, we don't parse a dict from file - but at the same time we generate config dicts for tests, and we want to test those using the code, putting them somewhere. to me, it feels right to think about how things might look like, and nudge the code towards that direction, rather than just nudging the code towards somehow running the tests. but of course the latter strategy might be just right, and we might end up with what we want the code to look like in the end...

> You would be either passing the filepath or the dict

Morten very passionately disagreed, arguing for a dictionary that truly reflects only the contents of the user config file. that necessitates to have something "come along" with the dictionary, to carry meta info, or context, which we need. Of course, one can just as well put everything in the dictionary. I dunno. I think separating it and having the separation clear and out in the open is a good idea, but there's a difference of opinions here...

> focus on fixing the tests and not adding/changing functionality
hmmm. yeah I mean if we can, that would be the best. I'm gonna have to think about that one. of course, I think that's what I originally tried to do - fix the tests, by stuffing the DEFINE keyword with predefineds in the generated config dicts, and using a test internal keyword. and we all know how much good that did me ; )

valentin-krasontovitsch replied:
Dan, for an alternative approach to passing the config file, see [my comment below](https://github.com/equinor/ert/pull/4215#discussion_r1021816113), or directly check out [Eivinds idea here](https://github.com/equinor/ert/pull/4215#issuecomment-1311225883)

eivindjahren replied:
@valentin-krasontovitsch I think going down the route of implementing `LibresFacade.from_dict` and moving everest is too soon right now. Also updating the `cli/gui/everest/semeio` to a new `__init__` is kind of unnecessarily messy as we would rather go straight to `LibresFacade.from_dict/LibresFacade.from_file`.

DanSava commented on src/ert/_c_wrappers/enkf/res_config.py:
```diff
@@ -174,6 +177,22 @@ def _log_config_content(self, config_content: ConfigContent) -> None:
         logger.info(tmp_dict)
 
+    @staticmethod
+    def _create_predefined_defines(
```
How does `_create_pre_defines` sound? In place of `_create_predefined_defines`.

valentin-krasontovitsch replied:
I'm still rooting for `predefineds`, but i'm not sure if it will stick ; )

will rename once we figure out where to move the method!

valentin-krasontovitsch replied:
it is done.

valentin-krasontovitsch commented on src/ert/_c_wrappers/enkf/res_config.py:
```diff
@@ -91,6 +91,7 @@ def __init__(
         user_config_file: Optional[str] = None,
         config: Optional[Dict[ConfigKeys, Any]] = None,
         config_dict: Optional[Dict[ConfigKeys, Any]] = None,
+        config_dict_source_file: Optional[str] = None,
```
We could also follow Eivind's suggestion ... somewhere ... and make the config dict a smarter thing. then it wouldn't be extending the number of arguments of res config, just changing the config dict and putting more information there. [there it is, a comment further up](https://github.com/equinor/ert/pull/4215#issuecomment-1311225883) - what do you think about that approach, Dan?

DanSava replied:
I think that approach is fine.

valentin-krasontovitsch replied:
winner winner chicken dinner! ( blast it, I agree with the approach as well, was just too lazy to actually implement it... :innocent: )

valentin-krasontovitsch commented:
okay i have now done the following:

- removed the test keyword `TEST_CONFIG_FILE_KEY`, it's gone
- stuck the predefined defines for test config dicts right in there under the `DEFINE_KEY`
- there is no more extension of res config's init arguments with `source_config_file`
- only make sure that the config file name in the dict and in the config file (in our config dict tests) are equal where it matters, namely whereever the substitution list is consideres
- dropped writing the config dir to the dict under the unused key `CONFIG_DIRECTORY`

waiting eagerly for feedback : )

valentin-krasontovitsch commented:
many thanks for the approval! :tada:
